opflexagent/gbp_ovs_agent.py

```diff
@@ -277,15 +277,70 @@ def port_bound(self, port, net_uuid,
+                mapping_copy['floating_ip'] = []
+                fip_by_fixed = {}
+                original_mac = mapping.get('mac_address') or port.vif_mac
+                # Prepare FIPs by fixed_ip
+                for fip in mapping.get('floating_ip', []):
+                    fip_by_fixed.setdefault(
+                        fip['fixed_ip_address'], []).append(fip)
+                # For the main MAC, set floating IP collection to all those
+                # FIPs pointing to the active Port fixed ips.
+                for fixed in fixed_ips:
+                    mapping_copy['floating_ip'].extend(
+                        fip_by_fixed.get(fixed['ip_address'], []))
+
+                # For the main MAC EP, set al the AAP with no mac address or
+                # MAC address equal to the original MAC.
+                for aap in mapping.get('allowed_address_pairs', []):
+                    if aap.get('active'):
+                        if not aap.get('mac_address') or aap.get(
+                                'mac_address') == original_mac:
+                            # Should go with the MAIN mac address EP file
+                            mapping_copy['allowed_address_pairs'].append(aap)
+                            # Also set the right floating IPs
+                            mapping_copy['floating_ip'].extend(
+                                fip_by_fixed.get(aap['ip_address'], []))
+                        else:
+                            # Store for future processing
+                            mac_active_aap.setdefault(
+                                aap['mac_address'], []).append(aap)
+                # Create mapping file for base MAC address
+                self.mapping_to_file(port, net_uuid, mapping_copy, fixed_ips,
                                      device_owner)
+                # Reset for AAP EP files
+                mapping_copy['allowed_address_pairs'] = []
+                mapping_copy['fixed_ips'] = []
+                mapping_copy['extra_ips'] = []
+                mapping_copy['subnets'] = []
+                mapping_copy['enable_dhcp_optimization'] = False
+                mapping_copy['enable_metadata_optimization'] = False
+                # Map to file based on the active AAP with a MAC address
+                for mac, aaps in mac_active_aap.iteritems():
+                    # Replace the MAC address with the new one
+                    mapping_copy['mac_address'] = mac
+                    # Extend the FIP list based on the allowed IPs
+                    mapping_copy['floating_ip'] = []
+                    mapping_copy['ip-address-mapping'] = []
+                    for aap in aaps:
+                        mapping_copy['floating_ip'].extend(fip_by_fixed.get(
+                            aap['ip_address'], []))
+                    # For this mac, set all the allowed address pairs.
+                    mapping_copy['allowed_address_pairs'] = aaps
```

Comment:
Not sure I understand your question. AAP just map to virtual-IP, and that is why they need to be promoted to fixed IP when a particular AAP is 'active'. If the AAP has a different MAC, then the IP-address appears in the fixed IPs for a different file; else it appears in the fixed IPs for the main file.

As for virtual-IP section in the additional EP file, my concern was that it may cause problems with managing lifecycle of flows added for the virtual-IPs. This may or may not be an issue, so I simply wanted to caution you.